### ISSUE TITLE
Should IP addresses be allowed for the client-facing server?

### ISSUE THREAD

**Issue by chris-wood** [tlswg/draft-ietf-tls-esni#424]
In #413, there's been a lot of discussion around how and whether or not IP addresses should be allowed as a reference identity of the client-facing server. This issue tracks sorting this particular question out. 

- @bemasc suggested [allowing the SNI be empty](https://github.com/tlswg/draft-ietf-tls-esni/issues/405#issuecomment-812088744) in the case the reference identity is an IP address. (Whether or not this address is in the ECHConfig is separate.)
- @martinthomson suggested [prohibiting IP addresses in the ECHConfig](https://github.com/tlswg/draft-ietf-tls-esni/pull/413#issuecomment-822089084). (Whether or not this allows the outer SNI to be empty is separate.)

So there seems to be two questions we need to sort out, in relation to #396:

1. Should the ECHConfig be allowed to supply IP addresses? And if so, how do client stacks validate them?
2. Should clients be allowed to validate servers using IP addresses as a reference identity? (If yes, then the ClientHelloOuter.SNI must be allowed to be empty/omitted.)

I think (2) should be 'yes', given that's currently possible today introduces no new complexity. I think it should be feasible to spell out the requirements to make (1) feasible, so I also think that should be 'yes'.

What do others think?

**Comment by bemasc** [comment]
Yes and yes.

RFC 8738 [fits the square peg in the round hole](https://tools.ietf.org/html/rfc8738#section-6) using reverse-IP notation, which seems like a halfway-reasonable option.  An "empty" public name seems a little more flexible (doesn't require servers to carry certificates for each other's IP addresses), and ought to be easy enough to support on clients.

**Comment by chris-wood** [comment]
@bemasc to clarify this:

> An "empty" public name seems a little more flexible (doesn't require servers to carry certificates for each other's IP addresses), and ought to be easy enough to support on clients.

You're suggesting that if ECHConfig is empty then the client should use the IP address of the client-facing server as the reference identity, right? That seems like a reasonable way around the language we might need to specify what is a valid address in ECHConfig. (If the reference identity is not empty, then it MUST be a name, otherwise it's the IP address, and clients should connect and validate based on that identity accordingly.)

**Comment by bemasc** [comment]
@chris-wood Correct.  @dmcardle previously raised a concern that this could complicate client implementations: the TLS stack can no longer establish a connection based solely on the ECHConfig and inner reference identity; it also needs the remote IP in some cases.  This is true, but I think the remote IP is essentially always "at hand" so perhaps it can be passed in unconditionally without too much difficulty.

**Comment by davidben** [comment]
I don't think that's true when the client is using a proxy, depending on how much of DNS is done on the client and how much is done on the proxy. If the client does the SVCB lookup itself but asks the proxy to look up TargetName, as the SVCB draft [itself suggests](https://github.com/MikeBishop/dns-alt-svc/blob/master/draft-ietf-dnsop-svcb-https.md#clients-using-a-proxy), the client won't have the remote IP of the client-facing server available.

(Are there weird networks that rewrite addresses at the DNS level? That would also break with this idea.)

**Comment by chris-wood** [comment]
To try and sharpen @davidben's point: will the client always have access to the client-facing server IP? I don't see how this _won't_ always be true, since the client connects to _something_, but maybe I'm misunderstanding the proxy scenario.

**Comment by bemasc** [comment]
@davidben is correct about the behavior when using a SOCKS5 or HTTP CONNECT proxy in domain-oriented mode.  The proxy performs A/AAAA resolution and TCP connection establishment, and does not provide this information back to the client, so the client doesn't know which server IP it is using.  That does suggest that "empty" `public_name` won't work.

**Comment by chris-wood** [comment]
Oh! I see. The proxy is the one establishing the transport connection. 👍

**Comment by davidben** [comment]
> domain-oriented mode

There's probably at least three modes here, if not more:

1. Proxy gets the origin DNS name and takes over the whole DNS lookup. Client doesn't do SVCB at all. The only way to salvage SVCB-based features is to invent a header for the proxy to tell the client what SVCB path it followed in the CONNECT response.
2. Client does SVCB lookup and gives TargetName to the proxy.
3. Client does the whole lookup and gives the IP address to the proxy.

The one that breaks empty names is (2). Though, given that proxies today are often used to access internal services (i.e. without the proxy, you may not even be able to resolve the name), I expect (1) to be the default for existing schemes. Whereas (3) might make sense for other kinds of proxy use cases. (I'm actually not entirely sure when you'd want (2) despite the spec recommendation... maybe if you're relying on the proxy to rewrite IPs, but the proxy needs to a DNS name to do so?)

**Comment by bemasc** [comment]
This is presumably not the right forum for this topic, but mode (2) is the one that gives you correct Happy Eyeballs, geo-DNS, etc.

**Comment by sftcd** [comment]
On 19/04/2021 15:58, David Benjamin wrote:
> (Are there weird networks that rewrite addresses at the DNS level?
> That would also break with this idea.)
I seem to recall some DNS IPv6 transition thing that maybe
has that property.

S.

**Comment by martinthomson** [comment]
Yes, there are transition technologies that rewrite addresses.  Clients get a v6 address because they only have v6, but the server address is v4.

As for the general proxy problem, there is no value in any design that does not involve the client having the ECHConfig.  It can't use the public IP (because it doesn't know it always), but it can still populate server_name according to the value of public_name.  I might have missed something there though.

From David's [options](https://github.com/tlswg/draft-ietf-tls-esni/issues/424#issuecomment-822608586), option 1 is a bit silly (it adds unnecessary latency), but it also gives you the best locality if your goal is to use the proxy egress as the client network location (option 2 as Ben [suggests](https://github.com/tlswg/draft-ietf-tls-esni/issues/424#issuecomment-822613850) only works if you care about the network location of the client, at which point you might as well not use a proxy).  The only difference between 2 and 3 there is whether the proxy needs to perform its own name resolution, which might differ from the client (and cause problems as a result; but that's mostly just a case of insisting that clients don't choose option 2).

I don't understand why it's so hard to require the use of a domain name.  It shouldn't be "[why not](https://github.com/tlswg/draft-ietf-tls-esni/pull/413#issuecomment-822488984)".  Adding other forms of reference identity adds complexity to what is already a morass of horrifying complexity, so we should insist on justification for every addition.  And I'm not seeing any concrete arguments in favour of the complex option.

So my answer is No to both questions.

**Comment by bemasc** [comment]
I don't agree with Martin's characterization of the proxy behaviors, but I don't think this is the place to discuss that topic anyway.

> I don't understand why it's so hard to require the use of a domain name.

To me, the main advantage of supporting IP reference identities is for origins that are not using a large-scale hosting provider, and thus cannot rely on a widely shared `public_name`.  Domain names are typically more closely linked to an organization than IP addresses (especially with VM hosting), and more expensive to change.

An IP address `public_name` ensures that the client does not populate the outer SNI, and supports a configuration where the default certificate only contains an IP address SAN.  This minimizes the information visible to an adversary unless they can also identify the service that runs on this IP address (e.g. by scanning the whole DNS).

An alternative to IP address in `public_name` would be a flag in the ECHConfig indicating that the specified reference identity is in the default certificate (which is implicitly true for IP reference identities).  This would allow clients to omit the SNI, improving small services' privacy against a passive adversary, but active adversaries could fetch the default certificate to learn the `public_name`.

**Comment by chris-wood** [comment]
For everyone's consideration, we now have #426 as an alternate solution to this problem (thanks to @davidben) which does two things: (1) avoids overloading a single field for two types of values (names and addresses), and (2) punts IP address validation to a future change. While I sympathize with Martin's view, it seems overly restrictive to just prohibit IP addresses outright. This change allows us to sort that out in a future change, and should make everyone happy. 

Thoughts?

**Comment by davidben** [comment]
Re #426, the idea is largely me going "oops, I'm sorry for the mess". :-) When I filed #405, I was mostly interested in the client validation. But most places with DNS strings have a semi-overlap with IP literal syntax, and RFC6066 already prohibits of IP literals. (Without declaring an actual syntax being prohibited!) So it seemed hard to resolve those questions without going one way or another on IPs. So I brought up that question too as a seemingly minor (hah) side question on the main issue.

That side question has turned out to be _way_ more messy than I anticipated. Since there isn't a strong need for IPs as public identities, but also unease to completely close that door, I propose we tweak the encoding so it's easy enough to add later (reuse the extension mechanism we're already putting in), but not actually bother in this draft.

**Comment by martinthomson** [comment]
Isn't it the case that the extension scheme would allow us to add IP addresses later, even without #426?

**Comment by chris-wood** [comment]
> Isn't it the case that the extension scheme would allow us to add IP addresses later, even without #426?

There's lots of different ways we can accomplish this, and it'll come down to personal taste. For example:

- Make public_name optional in the main ECHConfig (and waste the two byte length field), and allow an extension to specify the IP address.
- Move public_name to a mandatory extension (as 426 does), and allow future extensions to specify addresses override the name.
- etc.

Given the bikeshed that's happening here, it seems prudent to at least separate these things (names and addresses). If we can agree on that, then we just need to decide where we put those bits. #426 does the job one way. If someone wants to propose an alternative PR to put the bits somewhere else, please do so!

**Comment by martinthomson** [comment]
If we're really talking about a bikeshed, then the "no change" option works for me.  It's fewer bits on the wire (by 4), it is less complicated to enforce it being mandatory, and we already have it implemented.

I do think that public_name being optional has some impact here.  It means that you don't have a reference identity for fallback handling.  I think that's probably legitimate.  There might be cases where fallback is not needed.

**Comment by chris-wood** [comment]
> I do think that public_name being optional has some impact here. It means that you don't have a reference identity for fallback handling. I think that's probably legitimate. There might be cases where fallback is not needed.

Hmm... how do you see this working if a future extension wants to specify IP addresses as a fallback but the public_name is not allowed to be optional?

I agree that there will be cases where fallback is not needed. DNS-SD is one of them. And in that case, removing the public_name field (or allowing it to be empty, or whatever) will be needed. I think #426 addresses that pretty nicely.

**Comment by martinthomson** [comment]
Ahh, I misunderstood.  It's not that the extension is mandatory to use, but it is mandatory to understand when present.  Let's be really crisp about the words we use.

Of course, an empty value is an equally good way to indicate that the value isn't operative.  It's one byte in the less-common case to save 4 bytes in the common case.

The question occurs to me: is there any possibility that you could have both a public_name extension and this hypothetical ip_reference_identity extension?  What would that mean?

I want to be really open about my priors here: I think that extensibility along this axis is an anti-feature for ECH.  The mandatory-to-understand extensions are part of what really confirmed that for me.  I think that the current format would be improved by removing the extensions block.

**Comment by chris-wood** [comment]
> The question occurs to me: is there any possibility that you could have both a public_name extension and this hypothetical ip_reference_identity extension? What would that mean?

Unclear -- I think it'd be up to the hypothetical extension to define how those things interact. 

> I want to be really open about my priors here: I think that extensibility along this axis is an anti-feature for ECH. The mandatory-to-understand extensions are part of what really confirmed that for me. I think that the current format would be improved by removing the extensions block.

I hear you. If we didn't care about IP addresses for the fallback certificate, much of this would be simplified. (And admittedly the desire to remove the extensions block would increase for me, too.) I'm going to take this to the list, since we don't seem to be getting anywhere without consensus on that point.

**Comment by davidben** [comment]
> Ahh, I misunderstood. It's not that the extension is mandatory to use, but it is mandatory to understand when present. Let's be really crisp about the words we use.

The draft already defines a notion of "mandatory". Although #426 does both (let's call the other one "required"?), which I think is the minimal encoding change because...

> I do think that public_name being optional has some impact here. It means that you don't have a reference identity for fallback handling. I think that's probably legitimate. There might be cases where fallback is not needed.

I don't think we should open that can of worms. Even though not having a fallback flow is really the server's decision, I think the client has a stake in this too. It's useful to nudge the ecosystem towards things that work well. If a server deploys without fallback, underestimating the predictability of DNS caching, it will break rarely in some clients. That sort of thing is bad UX and increases the pressure for clients to adopt insecure fallbacks. Unless there is a clear need, I don't think we should allow that. (Otherwise not setting it up is deceptively attractive for a naive server since they make fewer decision.)

The extensibility story for ECH is a little funny. You send an ECHConfigList and the client ignores any ECHConfigs it can't understand. So if we say today's ECH clients require a public_name extension, future ECH clients can still support a public_ip extension that relaxes the public_name requirement. Today's ECH clients will just ignore the public_ip-using configs, which is the desired behavior.

But I should say I have no stake in this IP mess and do not care whether we allow those. Just pick something that everyone's happy with, please. #426 was an attempt to punt this out of the draft while keeping all the existing semantics unchanged (hence required extension; the correct reading is that it is *purely* an encoding change).

**Comment by martinthomson** [comment]
That point about making it hard NOT to provide a fallback makes me more inclined to reject #426.  If the effect is just a change in encoding, it's a less efficient encoding and one that is harder to validate properly.  If the effect is to make it optional under certain conditions, you have just argued effectively for why that has negative consequences.

**Comment by sftcd** [comment]
It's really a different point and maybe deserves it's own
issue but since I'm not logged into github and email works...

On 21/04/2021 06:04, David Benjamin wrote:
> The extensibility story for ECH is a little funny. You send an
> ECHConfigList and the client ignores any ECHConfigs it can't
> understand.
The "critical" bit in x.509 extensions IMO didn't work. IIRC
we really designed it for the basicConstraints extension, as
that had to be understood by relying parties, but the result
was that we more or less only had one shot at introducing any
critical extensions - unless I'm forgetting stuff (which is
entirely possible:-) pretty much all later attempts to add
new critical extensions failed as they'd break deployed
clients. We'd have been better off adding the is-a-ca flag
and anything else we knew to be critical the TBSCertificate.
(Again IIRC, we put both critical and non-critical stuff into
extensions because the *UniqueID fields in TBSCertificate
that were added in x.509v2 hadn't been useful so there was
a reluctance to put new fields in there at that point.)

I don't see why criticality is different for ECH extensions.
My bet is that any new critical extension will need a new TLS
extension code point once the first ECH RFC is issued.

So I'd remove the extensions fields we've defined and if not
then at least get rid of the idea of criticality.

Cheers,
S.
